File: firmware/src/modules/LEDs.py
from machine import Pin
import uasyncio as asyncio

import pins

LED_STATE_EXCELLENT = 0
LED_STATE_OK = 1
LED_STATE_WARN_WEAK = 2
LED_STATE_WARN_STRONG = 3
LED_STATE_ERROR = 4
LED_STATE_NONE = 5


class LEDs:
    def __init__(self):
        self.led_red = Pin(pins.LED_RED, mode=Pin.OUT, value=0)
        self.led_yellow = Pin(pins.LED_YELLOW, mode=Pin.OUT, value=0)
        self.led_green = Pin(pins.LED_GREEN, mode=Pin.OUT, value=0)

        self.blink_state_on = True

        self.led_state = LED_STATE_NONE

        self._update_leds()

    async def on_co2_changed(self, value):
        if value is None:
            self.led_state = LED_STATE_ERROR
        elif value >= 1450:
            self.led_state = LED_STATE_WARN_STRONG
        elif value >= 1000:
            self.led_state = LED_STATE_WARN_WEAK
        elif value >= 600:
            self.led_state = LED_STATE_OK
        elif value < 600:
            self.led_state = LED_STATE_EXCELLENT
        else:
            self.led_state = LED_STATE_NONE

        self._update_leds()

    async def run(self):
        while True:
            await asyncio.sleep_ms(600)
            self.blink_state_on = not self.blink_state_on
            self._update_leds()

    def _update_leds(self):
        green = 0
        yellow = 0
        red = 0

        if self.led_state == LED_STATE_ERROR:
            if self.blink_state_on:
                red = 1
                yellow = 1
                green = 1
        elif self.led_state == LED_STATE_WARN_STRONG:
            red = 1
        elif self.led_state == LED_STATE_WARN_WEAK:
            yellow = 1
        elif self.led_state == LED_STATE_OK:
            green = 1
        elif self.led_state == LED_STATE_EXCELLENT:
            if self.blink_state_on:
                green = 1
        else:
            red = 1
            yellow = 1
            green = 1

        self.led_green.value(green)
        self.led_yellow.value(yellow)
        self.led_red.value(red)

File: firmware_platformio/src/leds/LEDs.cpp
#include "LEDs.hpp"

#include "../Peripherals/pinout.hpp"
#include "../sensors/Sensors.hpp"

LEDs_t::LEDs_t()
    : blink_state{BLINK_PERIOD},
      blink_value{blink_state.get_blink_value()},
      co2_value{Sensors.get_co2_value()}
{
}

void LEDs_t::init()
{
    pinMode(PINS::LED_RED, OUTPUT);
    pinMode(PINS::LED_YELLOW, OUTPUT);
    pinMode(PINS::LED_GREEN, OUTPUT);

    update_leds();
}

void LEDs_t::update()
{
    blink_state.update();

    if (blink_value.new_value_available() ||
        co2_value.new_value_available())
    {
        update_leds();
    }
}

LEDs_t::LED_STATE LEDs_t::compute_led_state()
{
    if (!co2_value.is_valid())
        return LED_STATE_ERROR;

    const uint16_t value = co2_value.get();

    if (value >= 1450)
        return LED_STATE_WARN_STRONG;
    if (value >= 1000)
        return LED_STATE_WARN_WEAK;
    if (value >= 600)
        return LED_STATE_OK;
    if (value < 600)
        return LED_STATE_EXCELLENT;

    return LED_STATE_NONE;
}

void LEDs_t::update_leds()
{
    LED_STATE led_state = compute_led_state();

    uint8_t green = 0;
    uint8_t yellow = 0;
    uint8_t red = 0;

    if (led_state == LED_STATE_ERROR)
    {
        if (blink_value.get())
        {
            red = 1;
            yellow = 1;
            green = 1;
        }
    }
    else if (led_state == LED_STATE_WARN_STRONG)
    {
        red = 1;
    }
    else if (led_state == LED_STATE_WARN_WEAK)
    {
        yellow = 1;
    }
    else if (led_state == LED_STATE_OK)
    {
        green = 1;
    }
    else if (led_state == LED_STATE_EXCELLENT)
    {
        if (blink_value.get())
        {
            green = 1;
        }
    }
    else
    {
        red = 1;
        yellow = 1;
        green = 1;
    }

    digitalWrite(PINS::LED_GREEN, green);
    digitalWrite(PINS::LED_YELLOW, yellow);
    digitalWrite(PINS::LED_RED, red);
}

LEDs_t LEDs{};
